### PR TITLE
Fix: correctly associate SqlLoggedModelMetric with experiment_id

### PR DESCRIPTION
**Pull request by mcompen**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/mcompen/mlflow/pull/18382?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18382/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18382/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/18382/merge
```

</p>
</details>

### Related Issues/PRs
https://github.com/mlflow/mlflow/issues/16643#issuecomment-3409808800

 #xxx

### What changes are proposed in this pull request?
Get the experiment ID for a SqlLoggedModelMetric.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

1. Run local tracking server and set corresponding env vars (https://mlflow.org/docs/3.1.3/ml/tracking/tutorials/remote-server/)
1. Run script
```python
import mlflow
from mlflow.models.model import ModelInfo
import mlflow.sklearn
from sklearn.dummy import DummyClassifier
from mlflow.exceptions import MlflowException

try:
    mlflow.delete_experiment("0")
except MlflowException:
    print("Default experiment does not exist, continuing...")

try:
    mlflow.get_experiment("1")
except MlflowException:
    mlflow.create_experiment("test")

mlflow.set_experiment(experiment_id="1")
with mlflow.start_run() as run:
    X_train = [[0], [1], [2], [3]]
    y_train = [0, 1, 0, 1]
    model = DummyClassifier()
    model.fit(X_train, y_train)

    model_info: ModelInfo = mlflow.sklearn.log_model(model, name="dummy-classifier")
    mlflow.log_metric("f1", 0.7, model_id=model_info.model_id)
```
1. Observe experiment_id is set to 0 in postgres DB: `SELECT * FROM logged_model_metrics`
1. Delete all runs (or default experiment deletion will fail)
1. Delete default experiment `mlflow experiments delete -x 0`
1. Run `mlflow gc --tracking-uri http://127.0.0.1:5000  --backend-store-uri postgresql://user:password@localhost:5432/mlflowdb` to delete experiment from local postgres
1. Run script again: get ForeignKeyViolation error

Before fix:
image

After fix:
No ForeignKeyViolation error and correct experiment ID assocation
image

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

The SqlLoggedModelMetric associates all metrics with the default experiment (see https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/sqlalchemy_store.py#L902).
When the default experiment is deleted, this line results in ForeignKeyViolation (see https://github.com/mlflow/mlflow/issues/16643#issuecomment-3409808800).

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 1]
```diff
@@ -5739,6 +5739,42 @@ def test_log_logged_model_params(store: SqlAlchemyStore):
...
+
+    with store.ManagedSessionMaker() as session:
+        default_experiment = (
+            session.query(SqlExperiment)
+            .filter(SqlExperiment.experiment_id == int(store.DEFAULT_EXPERIMENT_ID))
+            .one_or_none()
+        )
+        if default_experiment:
+            session.delete(default_experiment)
+            session.commit()
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply tests/store/tracking/test_sqlalchemy_store.py]: +1

**Comment by mcompen** [comment]
@BenWilson2 I don't think that failing unit test is due to my changes. Could we rerun them?

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 2]
```diff
@@ -910,23 +909,34 @@ def _log_model_metrics(
...
+            sanitized_metrics.append((metric, value))
+
+        if not sanitized_metrics:
+            return
+
+        with self.ManagedSessionMaker() as session:
+            if experiment_id is None:
+                run = self._get_run(run_uuid=run_id, session=session)
+                self._check_run_is_active(run)
+                experiment_id = run.experiment_id
```
Comment (serena-ruan) [inline mlflow/store/tracking/sqlalchemy_store.py]: This fix makes sense, however there might be a performance issue when we log lots of metrics on the same run/experiment. We should avoid duplicate queries, is it possible to adjust so that experiment_id is always checked before passing here? cc @harupy

Comment (mcompen) [reply mlflow/store/tracking/sqlalchemy_store.py]: I was following the same pattern as for `_log_metrics`, `_log_params` and `_set_tags`. We can potentially reduce the number of calls by having all these methods return experiment ID s.t. it can be passed around when they are called one after another in `_log_batch`, but I don't think the performance increase is worth it. We probably still want to call `_get_run` in all these methods in any case to do the active run check.

Comment (serena-ruan) [reply mlflow/store/tracking/sqlalchemy_store.py]: Could we pass experiment_id from all upstream calls? Imagine you need to log hundreds of metrics, this change will cause significant performance issue by fetching runs in each metric call.
e.g. in fluent API `mlflow.log_metric` we already have the run object, we can just pass the experiment_id down

Comment (mcompen) [reply mlflow/store/tracking/sqlalchemy_store.py]: Or we add `experiment_id` to the Metric object? I see `run_id` is an attribute as well, but in practice it is always passed (or inferred) alongside the Metric. We could pass experiment_id alongside the Metrics object as well instead. Not sure if there is a broader philosophy, let me know which direction you prefer.

Comment (mcompen) [reply mlflow/store/tracking/sqlalchemy_store.py]: I made an attempt at including experiment_id from upstream alongside the Metric. Would need to have a second look for async calls. This does mean the active run check is skipped.

Comment (serena-ruan) [reply mlflow/store/tracking/sqlalchemy_store.py]: Sorry for the late reply! Looking at the changes it has pretty big impacts given it changes LogMetric proto, while this should only be used for metrics if there's a model linked to it. 
I would prefer adding a LogModelMetrics API for this purpose, so it's only invoked when model_id is not None and we fetch experiment_id once then pass it down. Any thoughts? @B-Step62 @harupy

Comment (mcompen) [reply mlflow/store/tracking/sqlalchemy_store.py]: I'm also curious to understand the decision to include experiment_id in the SqlLoggedModelMetric, but not in SqlMetric. Too many open ends for me 😄

Comment (serena-ruan) [reply mlflow/store/tracking/sqlalchemy_store.py]: > the decision to include experiment_id in the SqlLoggedModelMetric

Yeah I got more context on this, it's added to support filtering logged models within a set of experiments based on model metrics satisfying certain criteria.
Circling back on the fix, I think we can adapt your original fix for now but adjusting it to pass 'experiment_id' to _log_model_metrics from upstream? e.g. we fetch experiment_id linked to the run once within `log_metric` and `log_batch` APIs.

Comment (mcompen) [reply mlflow/store/tracking/sqlalchemy_store.py]: All right, updated it again.

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 3]
```diff
@@ -788,8 +787,13 @@ def _get_deleted_runs(self, older_than=0):
 
     def log_metric(self, run_id, metric):
         # simply call _log_metrics and let it handle the rest
+        with self.ManagedSessionMaker() as session:
+            run = self._get_run(run_uuid=run_id, session=session)
+            self._check_run_is_active(run)
+            experiment_id = run.experiment_id
```
Comment (serena-ruan) [inline mlflow/store/tracking/sqlalchemy_store.py]: Let's check `if metric.model_id` exists first and then get experiment_id? Since it's only needed for _log_model_metrics :)

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -7672,6 +7672,38 @@ def test_log_logged_model_params(store: SqlAlchemyStore):
...
+def test_log_model_metrics_use_run_experiment_id(store: SqlAlchemyStore):
+    exp_id = store.create_experiment(f"exp-{uuid.uuid4()}")
+    run = store.create_run(exp_id, "user", 0, [], "test_run")
+    model = store.create_logged_model(experiment_id=exp_id, source_run_id=run.info.run_id)
+
+    with store.ManagedSessionMaker() as session:
+        session.query(SqlExperiment).filter(
+            SqlExperiment.experiment_id == int(store.DEFAULT_EXPERIMENT_ID)
+        ).delete()
+        session.commit()
```
Comment (serena-ruan) [inline tests/store/tracking/test_sqlalchemy_store.py]: Why do we need this change?

Comment (mcompen) [reply tests/store/tracking/test_sqlalchemy_store.py]: The point is to check experiment association correctness. I wanted to particularly test this when no default experiment was present, but I have removed this part for simplicity.

**Review by serena-ruan — APPROVED** [review]
LGTM! Thanks so much for the fix and bear with my comments :)
